docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,335 @@
+                        self.log_debug(f"user '{user}' already included in group '{group.name}'")
+                    else:
+                        self.log_debug(f"user '{user}' not included in group '{group.name}'")
+                        try:
+                            self.log_info(f"adding user '{user}' to group '{group.name}'")
+                            Group.member_add(group.name, user)
+                            repairs += 1
+                        except GroupException as e:
+                            self.log_error(f"failed to add user '{user}' to group '{group.name}': {e}")
+                            failed_repairs += 1
+
+            elif attribute == "exclude":
+                users = members["exclude"]
+                for user in users:
+                    self.log_debug(f"user '{user}' should be excluded from group '{group.name}'")
+                    if user in group.members:
+                        self.log_debug(f"user '{user}' not excluded from group '{group.name}'")
+                        try:
+                            self.log_info(f"removing user '{user}' from group '{group.name}'")
+                            Group.member_remove(group.name, user)
+                            repairs += 1
+                        except GroupException as e:
+                            self.log_error(f"failed to remove user '{user}' from group '{group.name}': {e}")
+                            failed_repairs += 1
+                    else:
+                        self.log_debug(f"user '{user}' already excluded from group '{group.name}'")
+
+            elif attribute == "only":
+                users = members["only"]
+                self.log_debug(f"group '{group.name}' should only contain members {users}")
+                if (sorted(users) != sorted(group.members)):
+                    self.log_debug(f"group '{group.name}' does not only contain members {users}")
+                    try:
+                        self.log_info(f"setting members of group '{group.name}' to only users {users}")
+                        Group.members_set(group.name, users)
+                        repairs += 1
+                    except GroupException as e:
+                        self.log_error(f"failed to set members of group '{group.name}' to only users {users}: {e}")
+                        failed_repairs += 1
+                else:
+                    self.log_debug(f"group '{group.name}' does only contain members {users}")
+
+        return (failed_repairs, repairs)
+
+
+class GroupException(Exception):
+    def __init__(self, message):
+        self.message = message
+
+
+class Group:
+    def __init__(self, name, gid, members):
+        self.name = name
+        self.gid = gid
+        self.members = members
+
+    
+    @staticmethod
+    def lookup(group):
+        command = ["grep", f"^{group}", "/etc/group"]
+        process = Popen(command, stdout=PIPE, stderr=PIPE)
+        stdout, stderr = process.communicate()
```

Comment:
You can just do:
```
with open("/etc/group") as f:
  for line in f:
    if line.startswith(group):
      ...
```
No need to fork-exec *grep*.

Reply:
Agreed !